mailgun/route_handlers.py

```diff
@@ -33,45 +34,50 @@ def handle_mailing_list_email_route(request):
     :param request:
     :return JsonResponse:
     '''
-    logger.debug(u'Full mailgun post: %s', request.POST)
-
-    from_ = address.parse(request.POST.get('from'))
-    message_id = request.POST.get('Message-Id')
-    recipients = set(address.parse_list(request.POST.get('recipient')))
-    sender = address.parse(request.POST.get('sender'))
-    subject = request.POST.get('subject')
+    try:
+        logger.debug(u'Full mailgun post: %s', request.POST)
+
+        from_ = address.parse(request.POST.get('from'))
+        message_id = request.POST.get('Message-Id')
+        recipients = set(address.parse_list(request.POST.get('recipient')))
+        sender = address.parse(request.POST.get('sender'))
+        subject = request.POST.get('subject')
+
+        logger.info(u'Handling Mailgun mailing list email from %s (sender %s) to '
+                    u'%s, subject %s, message id %s',
+                    from_, sender, recipients, subject, message_id)
+
+        # short circuit if we detect a bounce loop
+        sender_address = sender.address.lower() if sender else ''
+        from_address = from_.address.lower() if from_ else ''
+        if settings.NO_REPLY_ADDRESS.lower() in (sender_address, from_address):
+            logger.error(u'Caught a bounce loop, dropping it. POST data:\n%s\n',
+                         json.dumps(request.POST))
+            return JsonResponse({'success': True})
+
+        for recipient in recipients:
+            # shortcut if we've already handled this message for this recipient
+            if message_id:
+                cache_key = (
+                    settings.CACHE_KEY_MESSAGE_HANDLED_BY_MESSAGE_ID_AND_RECIPIENT
+                        % (message_id, recipient))
+                if cache.get(cache_key):
+                    logger.warning(u'Message-Id %s was posted to the route handler '
+                                   u"for %s, but we've already handled that.  "
+                                   u'Skipping.', recipient, message_id)
+                    continue
+            _handle_recipient(request, recipient)
 
-    logger.info(u'Handling Mailgun mailing list email from %s (sender %s) to '
-                u'%s, subject %s, message id %s',
-                from_, sender, recipients, subject, message_id)
+    except:
+        logger.exception(
+            u'Unhandled exception; aborting. POST data:\n%s\n',
+            json.dumps(request.POST))
+        # tell Mailgun route forwarder to stop processing this message
+        return JsonResponse({'success': False}, status=406)
```

Comment:
Pretty sure we don't want to tell them to stop forwarding it to us.  It's entirely possible the failure could be caused by transient backend issues (ie. db down), and we want them to continue to retry.  I'd say return a 500.